vhost-device-spi/src/spi.rs

```diff
@@ -0,0 +1,1119 @@
+pub(crate) enum Error {
+    #[error("Ioctl command failed for {0} operation: {1}")]
+    IoctlFailure(&'static str, IoError),
+    #[error("Failed to open spi controller")]
+    DeviceOpenFailed,
+}
+
+/// Linux SPI definitions
+/// IOCTL commands, refer Linux's Documentation/spi/spidev.rst for further details.
+const _IOC_SIZEBITS: u32 = 14;
+const _IOC_SIZESHIFT: u32 = 16;
+const SPI_IOC_MESSAGE_BASE: u32 = 0x40006b00;
+
+ioctl_ior_nr!(SPI_IOC_RD_BITS_PER_WORD, 107, 3, u8);
+ioctl_iow_nr!(SPI_IOC_WR_BITS_PER_WORD, 107, 3, u8);
+ioctl_ior_nr!(SPI_IOC_RD_MAX_SPEED_HZ, 107, 4, u32);
+ioctl_iow_nr!(SPI_IOC_WR_MAX_SPEED_HZ, 107, 4, u32);
+ioctl_ior_nr!(SPI_IOC_RD_MODE32, 107, 5, u32);
+ioctl_iow_nr!(SPI_IOC_WR_MODE32, 107, 5, u32);
+
+// Corresponds to the SPI_IOC_MESSAGE macro in Linux
+fn spi_ioc_message(n: u32) -> u64 {
+    let mut size: u32 = 0;
+    if n * 32 < (1 << _IOC_SIZEBITS) {
+        size = n * 32;
+    }
+    (SPI_IOC_MESSAGE_BASE | (size << _IOC_SIZESHIFT)) as u64
+}
+
+/// Linux SPI mode, prefix with "LNX_"
+/// refer to Linux include/uapi/linux/spi/spi.h
+const LNX_SPI_CPHA: u32 = 1 << 0;
+const LNX_SPI_CPOL: u32 = 1 << 1;
+const LNX_SPI_CS_HIGH: u32 = 1 << 2;
+const LNX_SPI_LSB_FIRST: u32 = 1 << 3;
+const LNX_SPI_LOOP: u32 = 1 << 5;
+const LNX_SPI_TX_DUAL: u32 = 1 << 8;
+const LNX_SPI_TX_QUAD: u32 = 1 << 9;
+const LNX_SPI_TX_OCTAL: u32 = 1 << 13;
+const LNX_SPI_RX_DUAL: u32 = 1 << 10;
+const LNX_SPI_RX_QUAD: u32 = 1 << 11;
+const LNX_SPI_RX_OCTAL: u32 = 1 << 14;
+
+/// Config space supported mode mask
+const CONFIG_SPACE_TRANS_DUAL: u8 = 0x1;
+const CONFIG_SPACE_TRANS_QUAD: u8 = 0x2;
+const CONFIG_SPACE_TRANS_OCTAL: u8 = 0x4;
+const CONFIG_SPACE_CPHA_0: u32 = 0x1;
+const CONFIG_SPACE_CPHA_1: u32 = 0x2;
+const CONFIG_SPACE_CPOL_0: u32 = 0x4;
+const CONFIG_SPACE_CPOL_1: u32 = 0x8;
+const CONFIG_SPACE_CS_HIGH: u32 = 0x10;
+const CONFIG_SPACE_LSB: u32 = 0x20;
+const CONFIG_SPACE_LOOP: u32 = 0x40;
+
+/// Mode setting in Requests
+pub const SPI_CPHA: u32 = 1 << 0;
+pub const SPI_CPOL: u32 = 1 << 1;
+pub const SPI_CS_HIGH: u32 = 1 << 2;
+pub const SPI_LSB_FIRST: u32 = 1 << 3;
+pub const SPI_LOOP: u32 = 1 << 4;
+
+/// Copied (partially) from Linux's include/uapi/linux/spi/spidev.h
+///
+/// SpiIocTransfer - describes a single SPI transfer
+///
+/// @tx_buf: Holds pointer to userspace buffer with transmit data, or null.
+///
+/// @rx_buf: Holds pointer to userspace buffer for receive data, or null.
+///
+/// @len: Length of tx and rx buffers, in bytes.
+///
+/// @speed_hz: Temporary override of the device's bitrate.
+///
+/// @bits_per_word: Temporary override of the device's wordsize.
+///
+/// @delay_usecs: If nonzero, how long to delay after the last bit transfer
+///   before optionally deselecting the device before the next transfer.
+///
+/// @cs_change: True to deselect device before starting the next transfer.
+///
+/// @word_delay_usecs: If nonzero, how long to wait between words within one
```

Comment:
Some fields are missing here (e.g. tx_nbits, etc.)

Reply:
I will add the missing items.